libraries/core_libs/consensus/src/dag/sortition_params_manager.cpp

```diff
@@ -0,0 +1,176 @@
+
+uint16_t SortitionParamsChange::correctionPerPercent(SortitionParamsChange const& previous) const {
+  // make calculations on vrf upper threshold
+  const auto param_change = vrf_params.threshold_upper - previous.vrf_params.threshold_upper;
+  int32_t actual_efficiency_change = interval_efficiency - previous.interval_efficiency;
+  // to avoid division by zero
+  if (actual_efficiency_change == 0) {
+    actual_efficiency_change = 1;
+  }
+
+  auto correction_per_percent = std::abs(param_change * ONE_PERCENT / actual_efficiency_change);
+  // zero correction could lead to stuck of that protocol and give up with zero corrections
+  if (correction_per_percent == 0) {
+    correction_per_percent = 1;
+  }
+  // correction per percent shouldn't be bigger than percent of value
+  correction_per_percent = std::min(correction_per_percent, std::numeric_limits<uint16_t>::max() / 100);
+  return correction_per_percent;
+}
+
+bytes SortitionParamsChange::rlp() const {
+  dev::RLPStream s;
+  s.appendList(5);
+  s << vrf_params.threshold_lower;
+  s << vrf_params.threshold_upper;
+  s << period;
+  s << interval_efficiency;
+  s << actual_correction_per_percent;
+
+  return s.out();
+}
+
+SortitionParamsChange SortitionParamsChange::from_rlp(dev::RLP const& rlp) {
+  SortitionParamsChange p;
+
+  p.vrf_params.threshold_lower = rlp[0].toInt<uint16_t>();
+  p.vrf_params.threshold_upper = rlp[1].toInt<uint16_t>();
+  p.period = rlp[2].toInt<uint64_t>();
+  p.interval_efficiency = rlp[3].toInt<uint16_t>();
+  p.actual_correction_per_percent = rlp[4].toInt<uint16_t>();
+
+  return p;
+}
+
+SortitionParamsManager::SortitionParamsManager(const addr_t& node_addr, SortitionConfig sort_conf,
+                                               std::shared_ptr<DbStorage> db)
+    : config_(sort_conf), db_(std::move(db)) {
+  LOG_OBJECTS_CREATE("TARCAP");
+  // load cache values from db
+  params_changes_ = db_->getLastSortitionParams(config_.changes_count_for_average);
+  // restore VRF params from last change
+  if (!params_changes_.empty()) {
+    config_.vrf = params_changes_.back().vrf_params;
+  }
+
+  dag_efficiencies_.reserve(config_.computation_interval);
+  dag_efficiencies_ = db_->getLastIntervalEfficiencies(config_.computation_interval);
+}
+uint64_t SortitionParamsManager::currentProposalPeriod() const {
+  if (params_changes_.empty()) return 0;
+
+  return params_changes_.rbegin()->period;
+}
+
+SortitionParams SortitionParamsManager::getSortitionParams(std::optional<uint64_t> period) const {
+  if (!period) {
+    return config_;
+  }
+  auto p = config_;
+  for (auto prev = params_changes_.rbegin(), it = params_changes_.rbegin(); it != params_changes_.rend(); prev = it++) {
+    auto upper_bound = prev->period;
+    if (it->period == upper_bound) {
+      upper_bound = it->period + config_.computation_interval;
+    }
+    if (period >= it->period && period < upper_bound) {
+      p.vrf = it->vrf_params;
+    }
+  }
+  return std::move(p);
```

Comment:
no need to move on return

Reply:
I forgot why I did this, but during the build I got an error. So it is needed I guess 
```
taraxa-node/libraries/core_libs/consensus/src/dag/sortition_params_manager.cpp:92:10: error: local variable 'p' will be copied despite being returned by name [-Werror,-Wreturn-std-move]
  return p;
         ^
taraxa-node/libraries/core_libs/consensus/src/dag/sortition_params_manager.cpp:92:10: note: call 'std::move' explicitly to avoid copying
  return p;
         ^
         std::move(p)
```

Reply:
hmm, I think you've got this warning because p is reference.If it is so, then there is a bug as you are changing config_ basically

Reply:
can you try to replace `auto` for `SortitionParams`

Reply:
agree that std::move on return here is not ok

Reply:
Yeah it was a reference. But I didnt understand now why. I probably need to brush up on how the type inference for auto works